### PR TITLE
Create a generic controller

### PR DESCRIPTION
**Pull request by jsravn**
That watches ingress resources from Kubernetes. To be used by both the
dns and ingress controllers.

Also modify the ingress controller to use it.

### CODE REVIEW

**Comment by chbatey** [comment]
We now have three files called `controller.go`. Do we still need the utils one? Can the functions be moved to the generic one? Or at least the files named less generically for the specific controllers.

File: controller/ingress_entry.go [hunk 1]
```diff
@@ -28,8 +28,8 @@ type LoadBalancerEntry struct {
 }
 
 // FilterInvalidEntries returns a slice of all the valid LoadBalancer entries
```
Comment (chbatey) [inline controller/ingress_entry.go]: Ingresses rather than LoadBalancer. 

Do we want each Updater to validate rather than the controller or do you think this is applicable for all?

File: controller/ingress_entry.go [hunk 2]
```diff
@@ -60,14 +60,14 @@ func (entry LoadBalancerEntry) ValidateEntry() bool {
...
 	copy(sortedEntries, u.Entries)
 	sort.Sort(byName(sortedEntries))
-	return LoadBalancerUpdate{Entries: sortedEntries}
+	return IngressUpdate{Entries: sortedEntries}
 }
 
-type byName []LoadBalancerEntry
+type byName []IngressEntry
 
 func (a byName) Len() int           { return len(a) }
```
Comment (chbatey) [inline controller/ingress_entry.go]: Is it not clearer just to have the user call len?

**Comment by balooo** [comment]
`lb_test.go` should be `ingress_updater_test.go`

File: ingress/proxy.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package ingress
+
+import "github.com/sky-uk/feed/controller"
+
+// Proxy proxies traffic from external clients to the internal services in Kubernetes.
+type Proxy interface {
```
Comment (chbatey) [inline ingress/proxy.go]: Why do we need another interface for proxy? Is this just not an implementation of the Upater interface?

**Comment by jsravn** [comment]
> We now have three files called controller.go.

Is it a problem? I think in IntelliJ it shows the package as well, like `ingress/controller.go`. It's a bit annoying in intellij since the tab only shows the filename, but smurf naming controllers isn't great either. If we can think of a better name for controller.go then that's fine as well. I think the cmd one could probably change, or moved into controller.go.

**Comment by chbatey** [comment]
A few comments / questions. Assigning back.

File: controller/ingress_entry.go (hunk 2 above)
Comment (jsravn) [inline controller/ingress_entry.go]: This is required for sorting the entries. To sort something you have to implement `Len` and `Swap`. That was done in a separate PR when I ordered nginx entries.

File: controller/ingress_entry.go (hunk 1 above)
Comment (jsravn) [inline controller/ingress_entry.go]: For now it's applicable for all of them, although I think it may make more sense to put this logic inside the generic controller rather than standalone on the entry, since it is specific to the controller's logic.

File: ingress/proxy.go (hunk 3 above)
Comment (jsravn) [inline ingress/proxy.go]: Well we actually had `type Proxy controller.Updater`. I think it's nicer to make it explicitly for proxies rather than reuse updater, otherwise it gets pretty confusing having updaters calling updaters.

But the bool return value on Update doesn't make sense for the dns updater, so the interfaces don't actually match currently.
